feat: add allow-empty annotation.

jhoover4 opened the pull request:
### Description
PR to address feature change requested in #165.

Open to suggestions/review!

### Checklist
Please make sure that your PR fulfills the following requirements:
- [x] Reviewed the guidelines for contributing to this repository
- [x] The commit message follows the [Conventional Commits Guidelines](https://www.conventionalcommits.org/en/v1.0.0/#summary).
- [x] Tests for the changes have been updated
- [x] Docs have been added / updated

### Type of Change
<!-- Please check the one that applies to this PR using "x". -->
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] New tests
- [ ] Build/CI related changes
- [ ] Documentation content changes
- [ ] Other (please describe)

### Other information
<!-- Please add any additional information that would help reviewers evaluate your PR -->

werne2j commented:
@jhoover4 Looks like the tests are failing. Can you look into that? Thanks!

jhoover4 commented:
@werne2j whoops sorry about that! Looks like something got messed up in a merge. Its fixed now.

jkayani commented on pkg/kube/util_test.go:
```diff
@@ -329,6 +329,48 @@ func TestGenericReplacement_missingValue(t *testing.T) {
+		},
+	}
+
+	assertFailedReplacement(&dummyResource, &expected, t)
```
Why are we asserting failed here? If I read this right, the presence of the new annotation makes it so that `replacementErrors` is left empty if a placeholder can't be replaced. With that and the expected and actual data matching, this test seems wrong to me.

jhoover4 replied:
Ah I just copied a test above, I think both assertions would work cause we're just trying to assert that there are no errors. I can push a change making it `assertSuccessfulReplacement` if thats clearer? Or should I make a new assertion?

jkayani replied:
> I can push a change making it assertSuccessfulReplacement

Yes, this would be better and looks like you did that. Thanks!

jkayani commented:
This looks fine to me otherwise but I want to make sure we have a valid test - otherwise we'll have problems later on with catching regressions

werne2j commented:
@jhoover4 After looking it over more, not sure if the test is enough. Could you also add an example YAML as part of the end to end test. You can see the test i am referring to here https://github.com/IBM/argocd-vault-plugin/blob/main/cmd/generate_test.go#L66. 

Essentially, add a YAML to `fixtures/input/nonempty` and then add the expected outcome to `fixtures/output/all.yaml`. Will take a try or two to get the right order but i think it will be a good test.

jhoover4 commented:
Okay fair enough! I'll work on an e2e test tomorrow morning and get a new commit in.

jhoover4 commented:
Okay guys, I feel like I was kinda phoning it in before 😬. Took a closer look at this and I think the changes I just pushed are what my team really needs. We need the keys to be remove in a Secret or ConfigMap if they are missing. 

I've updated the annotation name to better reflect this and have restricted the annotation to Secrets and ConfigMaps only, since I believe any others would simply break the manifest.

Also added two e2e tests as requested, one testing the feature and another testing the error if any other manifest outside Secret or ConfigMap is tried with the annotation.

Let me know if all this makes sense or if I need to do some more work!

werne2j commented:
@jhoover4 Thanks for the update, we will take a look and discuss the proposed feature

jkayani reviewed (approved):
The feature and tests look to be implemented well. I'll leave it up to @werne2j to make the final decision. I appreciate you updating the docs as well.

werne2j commented:
@jhoover4 the only thing i'd like to see is example yamls added to the generate test. For reference: https://github.com/IBM/argocd-vault-plugin/pull/177#issuecomment-897097321 This will do a full test and not just start at the `replace`. Thanks!

jhoover4 commented:
> @jhoover4 the only thing i'd like to see is example yamls added to the generate test. For reference: [#177 (comment)](https://github.com/IBM/argocd-vault-plugin/pull/177#issuecomment-897097321) This will do a full test and not just start at the `replace`. Thanks!

Ah sorry I misunderstood. I think I got it now though! Let me know if what I added works.

werne2j reviewed (approved):
Looks good! Thanks @jhoover4
